### ISSUE TITLE
Surprising name resolution affecting SciJava LogService

### ISSUE THREAD

**Issue by sebi06** [jython/jython#70]
For details see the original report here:

[LogService issue with jython script in headless mode](https://github.com/fiji/fiji/issues/267)

and 

[https://forum.image.sc/t/logservice-issue-with-jython-slim-2-7-2-and-scripting-jython-1-0-0/40020/8](https://forum.image.sc/t/logservice-issue-with-jython-slim-2-7-2-and-scripting-jython-1-0-0/40020/8)

**The issue is really specific to jython-slim-2.7.2.jar. Removing that file and putting back jython-shaded-2.7.1.1.jar fixes the issue**

**Comment by ehrenfeu** [comment]
Dear Jython developers,

please note that this issue is **NOT** related to Fiji's / ImageJ's "headless" mode but also occurs when running the GUI.

My observations are that this is linked to running Jython in the SciJava context, where `jython-slim` seems to be getting confused with the private field [`info`][1] coming from [`AbstractRichPlugin`][2].

To reproduce:

* download a fresh [Fiji][3] (comes with `jython-slim-2.7.2.jar` included)
* extract and start it
* in the `File` menu select `New` and choose `Script...` (this will open the script editor)
* there, simply paste the example script below (mind the weird first line starting with `#@`)
* make sure to set the `Language` to `Python`
* then hit the `Run` button

```Python
#@LogService log

print type(log)
print type(log.error)
print type(log.warn)
print type(log.info)

log.setLevel(3)

log.log(1, "log1:error")
log.log(2, "log1:warn")
log.log(3, "log1:info")

log.error("log2:error")
log.warn("log2:warn")
log.info("log2:info")
```

As you can see in the output panel (switch between output and errors using the `Show Output` and `Show Errors` button) it reports type `org.scijava.log.StderrLogService` for the `log` object and type `instancemethod` for the `error()` and `warn()` methods, but confuses the `info` object to be of type `org.scijava.plugin.PluginInfo`.

Calling the `log()` method and specifying a log level is working fine ("`log1:...`" messages appear in the `Console` window).

Finally, the calls to `error()`, `warn()` are also working as expected, but the call to `info()` is resulting in a `TypeError: 'org.scijava.plugin.PluginInfo' object is not callable`.

As mentioned above, this does **NOT HAPPEN** when replacing `jython-slim-2.7.2.jar` with [`jython-shaded-2.7.1.1.jar`][4] in the `Fiji.app/jars/` directory - with 2.7.1.1 all calls are working fine.

[1]: https://github.com/scijava/scijava-common/blob/dcb568dcb9c3c215a508813e899ffc4f1bfff056/src/main/java/org/scijava/plugin/AbstractRichPlugin.java#L47-L48
[2]: https://github.com/scijava/scijava-common/blob/dcb568dcb9c3c215a508813e899ffc4f1bfff056/src/main/java/org/scijava/plugin/AbstractRichPlugin.java#L40-L41
[3]: https://fiji.sc
[4]: https://maven.scijava.org/service/local/repositories/central/content/org/scijava/jython-shaded/2.7.1.1/jython-shaded-2.7.1.1.jar

**Comment by jeff5** [comment]
The capability Jython has to ignore Java visibility has always seemed unwise to me. However, it is legitimate (I think) to have a public field and a public method both called `info`, and this would reproduce the circumstances independent of that feature.

When we build a type, and do not respect Java visibility, we follow a process that is materially different from the one normally taken (and prevailing in tests). The normal path seems to have had changes the other path hasn't. Maybe there are good reasons that the two can't use the same process. When I refactored, I preserved the different logic as I found it.

**Comment by jeff5** [comment]
This difficulty has been known about for a while:
https://github.com/jython/jython/blob/8c3734d1f09d0e2de8843c3d8f22b469b83e6860/Lib/test/test_java_visibility.py#L96-L98

There can only be one descriptor in the class `__dict__` by a given name, so it is difficult to see how both could be represented in that descriptor. However, we succeed in representing method polymorphism, so it is imaginable we could solve it. Can we give a reasonable meaning for access and assignment, however?

In the test suite, `org.python.tests.Visible.visibleInstance ` has signatures
```
public class Visible extends Invisible {
    public int visibleInstance = ...
    public int visibleInstance(int) ...
    public int visibleInstance(String) ...
    public int visibleInstance(int, String) ...
    public static int visibleInstance(String, String) ...
}
```
but only the *methods* are represented when we import the class
```
>>> for s in Visible.visibleInstance.argslist: print s.method
public int org.python.tests.Visible.visibleInstance(int)
public int org.python.tests.Visible.visibleInstance(java.lang.String)
public int org.python.tests.Visible.visibleInstance(int,java.lang.String)
public static int org.python.tests.Visible.visibleInstance(java.lang.String,java.lang.String)
```
These results are all with Jython respecting Java accessibility.

**Comment by jeff5** [comment]
Although there is a private field `info` in `org.scijava.plugin.PluginInfo`, this is not what causes the problem. (The comment above regarding the `respectJavaAccessibility` option is jumping to a conclusion.) Rather it is the dressing of `getInfo` and `setInfo` methods as Java Beam property `info`, which then takes precedence over the `info()` method.

I can reproduce the problem in Fiji (thanks @ehrenfeu for the clear instructions) and diagnose it with:
```
for cls in type(log).mro():
    print cls.__name__+".info =", cls.__dict__.get('info')
```
from which I get:
```
StderrLogService.info = None
AbstractLogService.info = None
AbstractService.info = None
AbstractRichPlugin.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x23>
AbstractContextual.info = None
LogService.info = <java function info 0x24>
SciJavaService.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x25>
Service.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x26>
RichPlugin.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x27>
SciJavaPlugin.info = None
Contextual.info = None
Prioritized.info = None
Comparable.info = None
HasPluginInfo.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x28>
Logged.info = None
Identifiable.info = None
Locatable.info = None
Versioned.info = None
Initializable.info = None
Disposable.info = None
Logger.info = <java function info 0x29>
Object.info = None
object.info = None
```
I can get essentially the same Jython 2.7.2 directly (needs `scijava-common-2.87.1.jar` on the class path). However, Jython 2.7.1 produces:
```
StderrLogService.info = None
AbstractLogService.info = None
LogService.info = <java function info 0x2>
SciJavaService.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x3>
AbstractService.info = None
Service.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x4>
AbstractRichPlugin.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x5>
RichPlugin.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x6>
SciJavaPlugin.info = None
AbstractContextual.info = None
Contextual.info = None
Prioritized.info = None
Comparable.info = None
HasPluginInfo.info = <beanProperty info type: org.scijava.plugin.PluginInfo 0x7>
Logged.info = None
Identifiable.info = None
Locatable.info = None
Versioned.info = None
Initializable.info = None
Disposable.info = None
Logger.info = <java function info 0x8>
Object.info = None
object.info = None
```

This reveals a change to the order of the MRO from 2.7.1 to 2.7.2, which is the solution to an [MRO conflict discovered in Eclipse (and IBM's MQQueue)](https://bugs.jython.org/issue2445). I can see that a lot of thought and investigation has gone into that issue, so I tend to think the change is well-founded.

However, I wonder if the `beanproperty` addition might somehow give way?

**Comment by jeff5** [comment]
In the Java API, `StderrLogService.info` refers to a method inherited from `LogService`. In Python we find it is shadowed by a property `HasPluginInfo.info` that in Java is a private field. Jython has not accessed the private field but synthesised this property from accessors `HasPluginInfo.getInfo` and `HasPluginInfo.setInfo`, because it interprets the Java Beans protocol.

The problem is not present for SciJava when using Jython 2.7.1 because it gets the MRO wrong, putting `LogService` first,. This was corrected in v2.7.2.

By "wrong" vs "corrected" I mean that if we set up a warren of Python classes, see below, where the inheritance mimics that of the Java classes involved here, the MRO computed by the reference implementation (CPython) agrees with 2.7.2 and not with 2.7.1, and an `info` `@property` from`HasPluginInfo.info` takes precedence. As noted above, the change was prompted by the [MRO conflict bjo 2445](https://bugs.jython.org/issue2445). This is correct, so it won't be reversed in 2.7.3.

I diagnose the root cause therefore to be an unfortunate side-effect of the automatic interpretation of get-set API as a bean property. This is a long-standing useful feature so isn't going to change in Jython 2.

However, the side-effect calls the bean idea into question for Jython 3. Maybe there's another way more like `@dataclass` attribute than `@property`.

For reference:
```
# python3_mro.py
# Class labyrinth comparabable to SciJava's

class Object: pass
class Logger:
    def info(msg): pass
class Disposable: pass
class Initializable: pass
class Versioned: pass
class Locatable: pass
class Identifiable: pass
class Logged: pass

class HasPluginInfo:
    @property
    def info(): return "my info"

##from dataclasses import dataclass
##@dataclass
##class HasPluginInfo:
##    info: object

class Comparable: pass
class Prioritized(Comparable): pass
class Contextual: pass

class AbstractContextual(Contextual): pass
class SciJavaPlugin: pass

class RichPlugin(SciJavaPlugin, Contextual, Prioritized,
	HasPluginInfo, Logged, Identifiable, Locatable, Versioned): pass
class Service(RichPlugin, Initializable, Disposable): pass

class AbstractRichPlugin(AbstractContextual, RichPlugin): pass
class AbstractService(AbstractRichPlugin, Service): pass
class SciJavaService(Service): pass
class LogService(SciJavaService, Logger): pass

class AbstractLogService(AbstractService, LogService): pass
class StderrLogService(AbstractLogService): pass

for cls in StderrLogService.mro():
    print(cls.__name__, ':', [c.__name__.split('.')[-1] for c in cls.mro()[1:]])

print()
descr = []
for cls in StderrLogService.__mro__:
    d = cls.__dict__.get('info')
    if d:
        descr.append(d)
        print(cls.__name__+".info =", d)
```

The output ends with the following "preference order":
```
HasPluginInfo.info = <property object at 0x000001AEA7400A90>
Logger.info = <function Logger.info at 0x000001AEA7405B40>
```
